Update mixer/v1 generated protos to use istio/gogo-genproto repo

douglas-reid opened the pull request:
This PR updates the API dependency on the googleapis protos. Instead of generating a non-existing mapping, the mapping is updated to the new istio/gogo-genprotos repository.

This PR is an attempt to make the repo (and API) more friendly towards go build tools. This approach is needed as Mixer uses a gogo-compiled version of the googleapis (and cannot use the `google.golang.com/genproto` version of these imports.

In addition, [go dep](https://github.com/golang/dep) support is added to the repo.

With this PR, it is now possible to:
```
$ go build ./...
```

Current plan: subsequent PRs will remove the bazel-based generation, opting for `go:generate` and gazelle-generated bazel files.

rshriram commented:
Don’t understand. Go build ./... will fail or succeed ? Your comments state it would fail.

rshriram commented on Gopkg.lock:
```diff
@@ -0,0 +1,45 @@
+  analyzer-version = 1
+  inputs-digest = "649c36bb02d48578a06a34bc83f740ef8ac1ce9389d93d8cce6e464baada888d"
+  solver-name = "gps-cdcl"
+  solver-version = 1
```
We don’t need this file. Auto generated and overwritten by dep.

costinm replied:
We certainly need this file as an artifact, included in the release/build, for reproduction.

Many projects check this file in - as a way to make sure developers can get a stable and tested set 
of deps, in case one of the deps gets broken.  It's a trade-off - it also means breakages in deps will
take longer to be fixed.

If we agree to checkin vendor/ - we can get the same effect.

I'm not actually sure if it's worth checking it in - but we must verify the file makes it into the build output.

rshriram commented on Gopkg.toml:
```diff
@@ -0,0 +1,38 @@
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/golang/protobuf"
```
Bad to pin sha’s? Pinning to master is going to make it difficult to invalidate cache. But if it happens rarely, then no issues

kyessenov replied:
@rshriram let's make it clear - we need both TOML and lock files. TOML has the constraints (use latest, or master), lock files has the pinned SHA for reproducibility. Once in a while, we can update lock files using TOML constraints, but that should be an independent and reviewed PR.

costinm replied:
See above, we need the lock file - but it doesn't have to be checked in. For example in android the lock file is just published with the release output. 

For protobuf - given that generated files are tied to a version, I think we should use the last release branch, not master, it's more likely to break in unexpected ways ( genfiles will likely be generated on proto changes - so changes
in protobuf lib would break us)

rshriram commented on mixer/v1/code.proto:
```diff
@@ -0,0 +1,185 @@
+// Copyright 2017 Google Inc.
```
Istio Authors
And in subsequent line, Google Inc

rshriram commented on mixer/v1/code.proto:
```diff
@@ -0,0 +1,185 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
And a line here stating that this has been copied from xyz repo for build purposes

rshriram commented on mixer/v1/error_details.proto:
```diff
@@ -0,0 +1,199 @@
+// Copyright 2017 Google Inc.
+//
```
Same comment as above.

rshriram commented on mixer/v1/status.proto:
```diff
@@ -0,0 +1,91 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
```
Ditto

wora commented:
I think it is a really bad idea to fork google.rpc package. If we can not make the build system work, it is more like our problem. I know Envoy is using google.rpc.Status directly.

Once Mixer starts to support remote adapters, we have to include third party proto files. We have to find a way to build them. It is not reasonable to fork other people's proto to make Mixer to build.

rshriram commented:
@wora this is an effort to get a pure Go build for istio/istio. In the interest of developer productivity/getting a sane build without complicated bazel dependencies, we have decided as a team to commit all generated protos into the repo (just like the way other projects do). This however results in a million line PR, when generating protos for google apis, when all we depend on is just the status proto (even there, just one or two things!).

I would rather check these in for the moment so that people can import istio/api as a pure go dependency in their projects.

kyessenov commented:
@wora there are no official gogo builds of googleapis, and Mixer made a decision to use gogo protobuf generator. We need a fork for gogo purposes, not only because we want to vendor protos.

kyessenov commented:
@douglas-reid have you tried putting protos into vendor together with generated files? We might want to preserve the import paths.

douglas-reid commented:
@kyessenov i have not tried to vendor the protos + generated files. that is certainly another reasonable approach. do we believe that is better (for reasons you mention: preserve import paths, clarity as to source) ?

I'm certainly open to it.

kyessenov commented:
That would satisfy @wora's concern about forking. I'm not sure how dep management tools would like that - would they try to flatten inner vendors and substitute with wrong repo? If we vendor, we probably need a special import path that is not `genproto` or `googleapis`, in which case, it's equivalent to just copying protos into the source tree.

douglas-reid commented:
@kyessenov lemme give it a try now.

douglas-reid commented:
@kyessenov experiment results:
- i can get the go build to work with vendoring of the protos (and generated files).
- i could not figure out how to easily get the cc_proto_libraries to build off of the generated files
- `dep ensure` either errors out attempting to overwrite the vendored files or deletes the vendored files if `ignored` is set for the package.

So, as I see it, we have a couple of options:
1) vendor, but don't support go dep in this repo
2) don't vendor, but find another place to put these protos.

Right now, I'm leaning towards creating a curated version of `google.golang.org/genprotos/googleapis` that only hosts the gogo* generated code for the protos we care about and linking to that repo with importmappings.

costinm commented:
Regarding 'forking' - golang seem to believe "A little copying is better than a little dependency" (in this 
case a huge dependency). The few protos we use are extremely unlikely to change, and arguably are general purpose enough, not tied to google. 

It would be better if gogo includes those basic protos and takes care of generating them, or if googleapis
splits them up ( and generates both gogo and std protos). The git repo including those files is huge and
probably takes the most time to checkout.

geeknoid commented:
Given the checkin to the new gogo-genproto repo, do we still need this PR?

rshriram commented:
closing as alternate solutions have been found.

rshriram commented:
oops.. seems its still valid.

douglas-reid commented:
PTAL.

I've updated this PR to simply update the mapping from the proto files to the gogo-genproto libraries.

I've also added dep support and verified `go build`-ability of the repo with these changes.

qiwzhang commented on Gopkg.toml:
```diff
@@ -0,0 +1,24 @@
+
+[[constraint]]
+  name = "github.com/gogo/protobuf"
+  version = "=0.5.0"
```
why there is a "=" prefix in the version?  is it a typo?

ldemailly replied:
it means exactly 0.5.0 though maybe it could be ~0.5 if they are good with compat ?

douglas-reid replied:
i opted for exact to match the gogo-genproto repo. i feel like we probably want to keep all repos in sync (though i don't have any real strong feelings here).

qiwzhang commented:
/approve
